server/ec2spotmanager/CloudProvider/GCECloudProvider.py

```diff
@@ -0,0 +1,376 @@
+        cluster = self._connect()
+
+        if pool_id is None:
+            libcloud_nodes = cluster.filter().has_labels([(SPOTMGR_TAG + "-PoolId").lower()]).nodes
+        else:
+            libcloud_nodes = cluster.filter().labels({(SPOTMGR_TAG + "-PoolId").lower(): str(pool_id)}).nodes
+
+        for node in libcloud_nodes:
+
+            if self.NODE_STATE_MAP[node.extra['status']] not in {INSTANCE_STATE['terminated']}:
+                instance_states[node.name] = {}
+                instance_states[node.name]['status'] = self.NODE_STATE_MAP[node.extra['status']]
+                instance_states[node.name]['tags'] = _LowercaseDict(node.extra['labels'])
+
+        return instance_states
+
+    def get_image(self, region, config):
+        assert region == "global", "Invalid region name for GCE: %s (only 'global' supported)" % (region,)
+        return config.gce_image_name
+
+    @staticmethod
+    def get_cores_per_instance():
+        return CORES_PER_INSTANCE
+
+    @staticmethod
+    def get_allowed_regions(_config):
+        return ["global"]
+
+    @staticmethod
+    def get_image_name(config):
+        return config.gce_image_name
+
+    @staticmethod
+    def get_instance_types(config):
+        return config.gce_machine_types
+
+    @staticmethod
+    def get_max_price(config):
+        return config.max_price
+
+    @staticmethod
+    def get_tags(config):
+        return config.instance_tags
+
+    @staticmethod
+    def get_name():
+        return 'GCE'
+
+    @staticmethod
+    def config_supported(config):
+        fields = ['gce_machine_types', 'max_price', 'gce_image_name', 'gce_container_name',
+                  'gce_disk_size']
+        return all(config.get(key) for key in fields)
+
+    def get_prices_per_region(self, region_name, instance_types=None):
+        # Pricing information is not perfect. The API Zones don't map to the data provided by Cloud Billing
+        # API. We usually get one price per zone (eg. us-east1), so we just assume that -a, -b, and -c have
+        # the same price. In some cases we do get details like "price in Tokyo" and "price in Japan"
+        # (both asia-northeast1), but we don't know which sub-zone those are, so we take the worst price and
+        # apply it to the whole zone to be conservative.
+        assert region_name == "global"
+
+        def get_price(sku):
+            assert len(sku["pricingInfo"]) == 1
+            expr = sku["pricingInfo"][0]["pricingExpression"]
+            assert len(expr["tieredRates"]) == 1, expr["tieredRates"]
+            rate = expr["tieredRates"][0]
+            unit = int(rate["unitPrice"]["units"])
+            nanos = rate["unitPrice"]["nanos"]
+            return ((unit + float(nanos) / 1e9), expr["usageUnit"])
+
+        zones = {}
+        service_id = "6F81-5844-456A"  # Compute Engine
+        params = {"key": settings.GCE_API_KEY}
+        data = requests.get("https://cloudbilling.googleapis.com/v1/services/" + service_id + "/skus",
```

Comment:
Maybe make a variable for the URL and reuse that down below so we don't duplicate the URL

Reply:
Thanks, I will.